packages/core/src/internal/data-grid/render/data-grid-render.lines.ts

```diff
@@ -146,6 +146,134 @@ export function overdrawStickyBoundaries(
+            maxY = Math.max(maxY, r.y + r.height + 1);
+        }
+    }
+
+    return { minX, maxX, minY, maxY };
+};
+
+export function drawExtraRowThemes(
+    ctx: CanvasRenderingContext2D,
+    effectiveCols: readonly MappedGridColumn[],
+    cellYOffset: number,
+    translateX: number,
+    translateY: number,
+    width: number,
+    height: number,
+    drawRegions: Rectangle[] | undefined,
+    totalHeaderHeight: number,
+    getRowHeight: (row: number) => number,
+    getRowThemeOverride: GetRowThemeCallback | undefined,
+    verticalBorder: (col: number) => boolean,
+    freezeTrailingRows: number,
+    rows: number,
+    theme: FullTheme
+) {
+    const bgCell = theme.bgCell;
+
+    const { minX, maxX, minY, maxY } = getMinMaxXY(drawRegions, width, height);
+
+    const toDraw: { x: number; y: number; w: number; h: number; color: string }[] = [];
+
+    const freezeY = height - getFreezeTrailingHeight(rows, freezeTrailingRows, getRowHeight);
+
+    // row overflow
+    let y = totalHeaderHeight;
+    let row = cellYOffset;
+    let y2 = 0;
+    while (y + translateY < freezeY) {
+        const ty = y + translateY;
+        const rh = getRowHeight(row);
+        if (ty >= minY && ty <= maxY - 1) {
+            const rowTheme = getRowThemeOverride?.(row);
+            const rowThemeBgCell = rowTheme?.bgCell;
+            const needDraw = rowThemeBgCell !== undefined && rowThemeBgCell !== bgCell && row >= rows;
+            if (needDraw) {
+                toDraw.push({
+                    x: minX,
+                    y: ty,
+                    w: maxX - minX,
+                    h: rh,
+                    color: rowThemeBgCell,
+                });
+            }
+        }
+
+        y += rh;
+        if (row < rows) y2 = y;
+        row++;
+    }
+
+    // column overflow
+    let x = 0;
+    const h = Math.min(freezeY, maxY) - y2;
+    if (h > 0) {
+        for (let index = 0; index < effectiveCols.length; index++) {
+            const c = effectiveCols[index];
+            if (c.width === 0) continue;
+            const tx = c.sticky ? x : x + translateX;
+            const colThemeBgCell = c.themeOverride?.bgCell;
+            if (
+                colThemeBgCell !== undefined &&
+                colThemeBgCell !== bgCell &&
+                tx >= minX &&
+                tx <= maxX &&
+                verticalBorder(index + 1)
+            ) {
+                toDraw.push({
+                    x: tx,
+                    y: y2,
+                    w: c.width,
+                    h,
+                    color: colThemeBgCell,
+                });
+            }
+
+            x += c.width;
+        }
+    }
+
+    if (toDraw.length === 0) return;
+
+    let color: string | undefined;
+    ctx.beginPath();
+    // render in reverse order because we computed and added the columns last, but they should actually be lower
+    // priority than the rows.
+    for (let i = toDraw.length - 1; i >= 0; i--) {
```

Comment:
The complexity here is just to reduce the number of draw calls. Fewer draw calls === faster